### ISSUE TITLE
Group types

### ISSUE THREAD

**Issue by RossNordby** [bepu/bepuphysics2#9]
The engine currently lacks nonconvex collidable types (and, at the time of writing, also [most convex types!](https://github.com/bepu/bepuphysics2/issues/8)).

To support nonconvex types, we should use a top level abstraction of group types. A group is a collidable bunch of shapes, but it prescribes no specific data structure or form. For practicality's sake, there will likely be two top level group types:

1. General shape groups containing any convex type, and
2. Meshes, or in other words, dedicated collections of triangles. Creating a special case for triangles is likely worth it due to their frequency of use and some special properties.

An ideal group type has a few functions:
1. Find all overlapping shapes in a bounding box for subpair dispatch.
2. Find results of other queries, like ray casts or sweep tests.
3. Ideally, boundary smoothing- a way to treat adjacent convex objects as a continuous surface with no 'bumps' at borders.

These properties do not require any specific data layout. Most likely, we will ship with:

1. Vertex/index list based meshes with scale/rotation/translation local transforms. They can use the Tree type to handle all the queries. Boundary smoothing is somewhat complicated, but is fundamentally similar to v1's implementation.
2. Terrain 'windows'. The idea here is that you can have multiple terrain collidables, each covering a portion of the underlying data set while still supporting some form of boundary smoothing at the collidable boundary due to the shared underlying data. The data source can be generic, so these windows could cover a procedural infinite terrain or anything else. To avoid virtual dispatches at runtime, specific generic versions will have to be registered in the collision task registry. We may want to just show examples of custom data sources in the demo rather than shipping in-engine registered versions.
3. Compounds. Similar to v1's CompoundShape, being able to stick a bunch of other shapes together is important. No compound nesting though- that was a pretty goofy feature. It may be possible to efficiently smooth boundaries within a compound- if all shapes come with a speedy 'get shortest vector to shape surface from point', you could check to see if the contacts you've found are obstructed by adjacent shapes in the group. That's pretty experimental- probably should wait on implementing that.

But there's nothing stopping you from doing some more interesting things. For example:
1. Voxel regions. Similar idea to terrain windows, except instead of triangles, you create boxes. Hardest part is still just the boundary smoothing.
2. Dynamically triangulated isosurfaces. With appropriate acceleration structures, you could go directly from something like a distance field to a rigid surface.

An example simple voxel region in the demos could be useful for users learning how to customize the engine.

**Comment by RossNordby** [comment]
While compounds and meshes are now available, there is still no unifying abstraction for them. There's quite a bit of reusable stuff for types which can be considered a Compound or a Mesh in an abstract sense, but there's one glaring hole that complicates adding something like a boundary smoothed voxel collidable:

Continuation types are currently hardcoded and dispatched with a switch. Right now, there exist NonconvexReductions, MeshReductions, and CompoundMeshReductions, but none of those could be used to smooth contacts in a hypothetical custom voxel collidable. 

In the long run, to support that level of extensibility, it's likely that we'll have to move over to a registry-with-ids model for continuations. Just like you can currently register a custom collision task for a particular type pair, but for continuations.

Arguably, having continuations as distinct entities (rather than just being additional collision tasks) is a bit weird. The only blocker there is that collision tasks tend to work by deferred execution, and continuations execute as soon as the data is available. Collision tasks require a flush to push any pending executions for this reason. Currently, compound tasks are allowed to generate subtasks, so this isn't anything too strange- but the task ordering logic only considers direct subtask generation. If more complex trees of tasks are allowed, the task DAG would need to be flattened for single pass flushing like we did originally.

**Comment by RossNordby** [comment]
Closing in favor of more specific issues later.
